userdata.tpl

```diff
@@ -2,9 +2,134 @@ MIME-Version: 1.0
+
+  # Loop over each of the lines and add each key value pair to the associate array
+  # Example output:
+  # [--log-file]="/tmp/kubelet.log"
+  # [--hostname-override]="dedicated-node-az1"
+  while IFS=% read -r key value; do
+    NEW_EXTRA_ARGS[$key]="$value"
+  done < <(prepare_lines "$KUBELET_NEW_ARGS")
+
+  for KEY in "$${!NEW_EXTRA_ARGS[@]}"; do
+    # If there is existing values, add a comma before adding the new values
+    test -n "$${EXTRA_ARGS[$KEY]}" && EXTRA_ARGS[$KEY]+=","
+    EXTRA_ARGS[$KEY]+="$${NEW_EXTRA_ARGS[$KEY]}"
+  done
+fi
+
+# Additional taints from Terraform, separated by commas
+# Example:
+# dedicated=new-test:NoSchedule
+NEW_TAINTS='${node_taints}'
+
+# Additional labels from Terraform, separated by commas
+# Example:
+# new.label=new
+NEW_LABELS='${node_labels}'
+
+# Add the new taints to the associate array, if defined
+# Example output:
+# [--node-labels]="eks.amazonaws.com/nodegroup=test-mng,other.label=other"
+# [--register-with-taints]="dedicated=test:NoSchedule,dedicated=new-test:NoSchedule"
+if [[ -n "$NEW_TAINTS" ]]; then
+  # If there is existing taints, add a comma before adding the new taints
+  test -n "$${EXTRA_ARGS['--register-with-taints']}" && EXTRA_ARGS['--register-with-taints']+=","
+  EXTRA_ARGS['--register-with-taints']+="$NEW_TAINTS"
+fi
+
+# Add the new labels to the associate array, if defined
+# Example output:
+# [--node-labels]="eks.amazonaws.com/nodegroup=test-mng,other.label=other,new.label=new"
+# [--register-with-taints]="dedicated=test:NoSchedule,dedicated=new-test:NoSchedule"
+if [[ -n "$NEW_LABELS" ]]; then
+  # If there is existing labels, add a comma before adding the new labels
+  test -n "$${EXTRA_ARGS['--node-labels']}" && EXTRA_ARGS['--node-labels']+=","
+  EXTRA_ARGS['--node-labels']+="$NEW_LABELS"
+fi
+
+# Build string from the joined arguments to add to the output file
+# Example:
+# --log-file=/tmp/kubelet.log --hostname-override=dedicated-node-az1 --node-labels=eks.amazonaws.com/nodegroup=test-mng,other.label=other,new.label=new --register-with-taints=dedicated=test:NoSchedule,dedicated=new-test:NoSchedule
+NEW_KUBELET_EXTRA_ARGS=""
+for i in "$${!EXTRA_ARGS[@]}";do
+  NEW_KUBELET_EXTRA_ARGS+=$(printf "%s=%s " "$i" "$${EXTRA_ARGS[$i]}")
+done
+
+# Remove the trailing space
+NEW_KUBELET_EXTRA_ARGS=$(echo "$NEW_KUBELET_EXTRA_ARGS" | xargs)
+
+# Write the final output file
+# Example:
+# [Service]
+# Environment='KUBELET_EXTRA_ARGS=--log-file=/tmp/kubelet.log --hostname-override=dedicated-node-az1 --node-labels=eks.amazonaws.com/nodegroup=test-mng,other.label=other,new.label=new --register-with-taints=dedicated=test:NoSchedule,dedicated=new-test:NoSchedule'
+cat <<CONF > /etc/systemd/system/kubelet.service.d/30-kubelet-extra-args.conf
+[Service]
+Environment='KUBELET_EXTRA_ARGS=$${NEW_KUBELET_EXTRA_ARGS}'
+CONF
+EOF
+
+# Execute the hook script before starting the services
+sed -i "/^systemctl daemon-reload\$/i . /etc/eks/extra_args.sh" /etc/eks/bootstrap.sh
+--//
```

Comment:
While I applaud the effort to be comprehensive, this still seems to me a lot more complicated and brittle than it needs to be. I do not want to be supporting this a year from now.